index.js

```diff
@@ -156,24 +135,64 @@ const useUint8Array = chunk => new Uint8Array(chunk);
 
 const useUint8ArrayWithOffset = chunk => new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength);
 
-const getContentsAsArrayBuffer = (chunks, textDecoder, length) => {
-	const contents = new Uint8Array(length);
+// `contents` is an increasingly growing `Uint8Array`.
+const addArrayBufferChunk = (convertedChunk, contents, length, previousLength) => {
+	const newContents = hasArrayBufferResize() ? resizeArrayBufferFast(contents, length) : resizeArrayBuffer(contents, length);
+	newContents.set(convertedChunk, previousLength);
+	return newContents;
+};
+
+// Without `ArrayBuffer.resize()`, `contents` size is always a power of 2.
+// This means its last bytes are zeroes (not stream data), which need to be
+// trimmed at the end with `ArrayBuffer.slice()`.
+const resizeArrayBuffer = (contents, length) => {
+	if (length <= contents.length) {
+		return contents;
+	}
+
+	const newContents = new Uint8Array(getNewContentsLength(length));
+	newContents.set(contents, 0);
+	return newContents;
+};
 
-	let offset = 0;
-	for (const chunk of chunks) {
-		contents.set(chunk, offset);
-		offset += chunk.length;
+// With `ArrayBuffer.resize()`, `contents` size matches exactly the size of
+// the stream data. It does not include extraneous zeroes to trim at the end.
+// The underlying `ArrayBuffer` does allocate a number of bytes that is a power
+// of 2, but those bytes are only visible after calling `ArrayBuffer.resize()`.
+const resizeArrayBufferFast = (contents, length) => {
```

Comment:
I would name this `resizeArrayBuffer` and the other one `resizeArrayBufferSlow` as this one is the one we will keep in the future. It makes it slightly easier to remove the old one when we target Node.js 20.

Reply:
Done in bcb87e7c80af79a50703134afa8001167b4fe8fb :+1: